p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,351 @@
+package nodes
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/lazyledger/nmt"
+
+	"github.com/ipfs/go-ipfs/core/coredag"
+	"github.com/ipfs/go-ipfs/plugin"
+	format "github.com/ipfs/go-ipld-format"
+	node "github.com/ipfs/go-ipld-format"
+	mh "github.com/multiformats/go-multihash"
+)
+
+const (
+	// NMT is the codec used for this plugin.
+	// 0x77 seems to be free:
+	// https://github.com/multiformats/multicodec/blob/master/table.csv
+	NMT = 0x1077
```

Comment:
You can have two codecs for both `nmtNode` and `nmtLeafNode` what would allow you not to store `nmt.NodePrefix` and `nmt.LeafPrefix` inside node's `RawData()` to save few bytes on every node. Furthermore, that would allow you to understand what type of node your `Link` is pointing to without actually having the node:

D - downloaded (saved in local IPFS repo)
ND - not downloaded (somewhere on the network)
-> - link points to

   | D |       | ND |
1. nmtNode -> nmtLeafNode
   | D |       | ND |      | ND |
2. nmtNode -> nmtNode -> nmntLeafNode

In both cases, you would understand the Node type without even having the linked node by checking its CID codec and save a byte for each :)

Reply:
However, that will require you to move from simple hashes in NMTHasher to use CIDs instead, as it sums prefixes as well, or to create a custom CIDHasher for nmt

Reply:
However, that will require you to move from simple hashes in NMTHasher to use CIDs instead, as it sums prefixes as well, or to create a custom CIDHasher for nmt

Reply:
> that will require you to move from simple hashes in NMTHasher to use CIDs instead

You mean directly in the tree implementation? 

> or to create a custom CIDHasher for nmt

The hasher is pluggable in the nmt implemtation. That sounds like that should work.

Reply:
> You can have two codecs for both nmtNode and nmtLeafNode what would allow you not to store nmt.NodePrefix and nmt.LeafPrefix inside node's RawData() to save few bytes on every node. Furthermore, that would allow you to understand what type of node your Link is pointing to without actually having the node:

@Wondertan wouldn't having two custom multihashes have the same result? That way the CIDs contain contain the information if they correspond to a leaf or an inner node (without having to download the actual rawData?

Reply:
Yes, that should have the same effect, as CID wraps MH.

Reply:
Pls clarify what you mean by
> requirement per se

Reply:
> That single byte is not a problem and it simplifies the current implementation a lot.
Agree with you, though I would consider possible migration issues that can appear in the future. For me, it looks like spending more time now, would save lots of time later on.

Reply:
It's so early in this project that migration won't be an issue. We will likely do a lot more breaking changes in the coming months. I hope everything to stabilize by end of this year only. Of course, we shouldn't introduce any hard to overcome technical debt in the meantime but I don't think this is the case here as it is a contained part of the code that cares about this (namely ipld only).

Also, what would you store in an (ipld) block for an inner node instead? Currently, it's the raw data of the children prefixed by a single byte. If you don't this, how would reading a block and creating an ipld Node including its links work? Storing the CIDs to create the Links of that node is even more bytes (implicitly including a byte to differentiate between inner / leaf nodes for the children). 

> >Pls clarify what you mean by

> requirement per se

In LL, the trees are always fully balanced binary trees and nodes that are sampling or reading from them know the tree depth up front. This means they know at which depth to expect a leaf. 

Does that make sense?